File: backend/internal/websocket/lobby_session_registry.go
package websocket

import (
	"sync"

	"github.com/google/uuid"
	"github.com/gorilla/websocket"
)

type Session struct {
	ID   string
	Conn *websocket.Conn
	mu   sync.Mutex
}

type LobbySessionRegistry struct {
	sessionToUser map[string]uuid.UUID
	sessions      map[string]*Session
	mu            sync.RWMutex
}

func NewLobbySessionRegistry() *LobbySessionRegistry {
	return &LobbySessionRegistry{
		sessionToUser: make(map[string]uuid.UUID),
		sessions:      make(map[string]*Session),
	}
}

func (r *LobbySessionRegistry) RegisterSession(session *Session) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.sessions[session.ID] = session
}

func (r *LobbySessionRegistry) UnregisterSession(sessionID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.sessions, sessionID)
	delete(r.sessionToUser, sessionID)
}

func (r *LobbySessionRegistry) BindUser(sessionID string, userID uuid.UUID) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.sessionToUser[sessionID] = userID
}

func (r *LobbySessionRegistry) UnbindUser(sessionID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.sessionToUser, sessionID)
}

func (r *LobbySessionRegistry) FindUserID(sessionID string) (uuid.UUID, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	userID, ok := r.sessionToUser[sessionID]
	return userID, ok
}

func (r *LobbySessionRegistry) GetAllSessions() []*Session {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result := make([]*Session, 0, len(r.sessions))
	for _, session := range r.sessions {
		result = append(result, session)
	}
	return result
}

func (s *Session) Lock() {
	s.mu.Lock()
}

func (s *Session) Unlock() {
	s.mu.Unlock()
}

File: backend_legacy/src/main/java/com/dominoes/lobby/websocket/LobbySessionRegistry.java
package com.dominoes.lobby.websocket;

import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class LobbySessionRegistry {

    private final Map<String, UUID> sessionToUser = new ConcurrentHashMap<>();
    private final Map<String, WebSocketSession> sessions = new ConcurrentHashMap<>();

    public void registerSession(WebSocketSession session) {
        sessions.put(session.getId(), session);
    }

    public void unregisterSession(String sessionId) {
        sessions.remove(sessionId);
        sessionToUser.remove(sessionId);
    }

    public void bindUser(String sessionId, UUID userId) {
        sessionToUser.put(sessionId, userId);
    }

    public void unbindUser(String sessionId) {
        sessionToUser.remove(sessionId);
    }

    public Optional<UUID> findUserId(String sessionId) {
        return Optional.ofNullable(sessionToUser.get(sessionId));
    }

    public Set<WebSocketSession> getAllSessions() {
        return Set.copyOf(sessions.values());
    }
}
